### PR TITLE
Implements SA1109 (+tests)

### LINKED ISSUE #43: SA1109: BlockStatementsMustNotContainEmbeddedRegions
http://www.stylecop.com/docs/SA1109.html

### PR DESCRIPTION
**Pull request by vweijsters**
Fixes #43

The tests for the code fix are there, but they are currently disabled as the code fix implementation is still a dummy.

### CODE REVIEW

**Comment by pdelvo** [comment]
Why dont you just use the RemoveRegionCodeFixProvider? Just add the diagnostic id to the set of supported diagnostic ids

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1109BlockStatementsMustNotContainEmbeddedRegions.cs [hunk 1]
```diff
@@ -23,35 +30,119 @@
...
+        private static void CheckTrivia(SyntaxNodeAnalysisContext context, IEnumerable<SyntaxTrivia> triviaList)
         {
-            // TODO: Implement analysis
+            var violatingDirective = triviaList.FirstOrDefault(t => t.IsKind(SyntaxKind.RegionDirectiveTrivia) || t.IsKind(SyntaxKind.EndRegionDirectiveTrivia));
+
+            Location location;
+            switch (violatingDirective.Kind())
+            {
+                case SyntaxKind.RegionDirectiveTrivia:
+                    location = Location.Create(context.Node.SyntaxTree, TextSpan.FromBounds(violatingDirective.SpanStart, violatingDirective.SpanStart + 7));
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1109BlockStatementsMustNotContainEmbeddedRegions.cs]: What is magic number 7?

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1109BlockStatementsMustNotContainEmbeddedRegions.cs [hunk 2]
```diff
@@ -23,35 +30,119 @@
...
+                    context.ReportDiagnostic(Diagnostic.Create(SA1109Descriptor, location));
+                    break;
+                case SyntaxKind.EndRegionDirectiveTrivia:
+                    location = Location.Create(context.Node.SyntaxTree, TextSpan.FromBounds(violatingDirective.SpanStart, violatingDirective.SpanStart + 10));
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1109BlockStatementsMustNotContainEmbeddedRegions.cs]: Ditto 10

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1109CodeFixProvider.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        /// <inheritdoc/>
+        public override FixAllProvider GetFixAllProvider()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        /// <inheritdoc/>
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            foreach (var diagnostic in context.Diagnostics.Where(d => this.FixableDiagnosticIds.Contains(d.Id)))
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1109CodeFixProvider.cs]: delegate and closure allocation at simple task :cry:

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1109BlockStatementsMustNotContainEmbeddedRegions.cs (hunk 1 above)
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1109BlockStatementsMustNotContainEmbeddedRegions.cs]: `"#region".Length` seems like the obvious guess.

**Comment by sharwell** [comment]
> Why dont you just use the RemoveRegionCodeFixProvider? Just add the diagnostic id to the set of supported diagnostic ids

Based on the tests, the code fix moves the region directives inside the curly braces instead of actually removing them. I wouldn't be opposed to changing the tests to expect removal.

**Comment by sharwell** [comment]
@vweijsters @pdelvo @Maxwe11 : #998 *directly* impacts this pull request. Let me know what you think.

**Comment by sharwell** [comment]
Blocked on #998.

**Comment by sharwell** [comment]
Closing as **wontfix** based on #998.

**Pull request closed without merging**